### PR TITLE
[pulsar-broker] Support of TopicLifecyclePolicies REST handler.

### PR DESCRIPTION
**Pull request by ciaocloud**
Originally authored by @merlimat, fixed for compatibility with 2.10 branch. (Note that with the first commit only, it would adapt to 2.9 branch)

### Motivation

*Support of TopicLifecyclePolicies REST handler.*

### Modifications
The added endpoint `TopicLifecycle` can be used during tenant creation time.
Both `InactiveTopicPolicies` and `TopicLifecycle` are fields declared in `Policies`. The `InactiveTopicPolicies` has a boolean value `deleteWhileInactive` which determines whether topics should be automatically deleted when inactive. We ask it to be same as that defined in `TopicLifecycle`, while `TopicLifecycle` additionally defines whether topics should be automatically created when there exists producer/consumer connection (which is oppose to being created explicitly with API).

Specifically, our changes include:
- defined TopicLifecyclePolicies class
- declared in Policies
- reset new inactiveTopicPolicies with topicLifecycle info if one exists while initializing and updating PersistentTopic
- added setter methods in broker-admin Namespaces.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change is a trivial rework / code cleanup without any test coverage.

### Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

- [x] `no-need-doc`

### CODE REVIEW

**Comment by Anonymitaet** [comment]
Hi @ciaocloud is this a code cleanup work and do not need to update docs?

when submitting a PR, can you help provide a doc label (tick the box) in the [PR template which contains info about doc](https://github.com/apache/pulsar/blob/master/.github/PULL_REQUEST_TEMPLATE.md#documentation)? This helps others know more about the changes. Thanks
Documentation

**Review by Jason918 — COMMENTED** [review]
I am confused about the relation between `InactiveTopicPolicies` and `TopicLifecycle`. Can you bring more details?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 1]
```diff
@@ -2485,6 +2493,13 @@ private boolean shouldTopicBeRetained() {
         //If the topic-level policy already exists, the namespace-level policy cannot override the topic-level policy.
         Optional<TopicPolicies> topicPolicies = getTopicPolicies();
 
+        if (data.topicLifecycle != null) {
+            // reset new inactiveTopicPolicies with topicLifecycle info if one exists
+            this.topicPolicies.getInactiveTopicPolicies().updateBrokerValue(
```
[the remark this answers is not in the document]
Comment (ciaocloud) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2498]: we wished to set the `inactiveTopicPolicies` as the new one, isn't it the right way to do it with your recent changes? in 2.9, the `inactiveTopicPolicies` was a member field in PersistentTopic.

[the remark this answers is not in the document]
Comment (ciaocloud) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2498]: o.k., we'll need to use `updateNamespaceValue` to reset the `inactiveTopicPolicies`, right?

File: pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/TopicLifecyclePolicies.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+public class TopicLifecyclePolicies {
+
+    /**
+     * Whether topics should be automatically created when producers/consumers connect, or rather explicitly created
+     * through admin API.
+     */
+    private boolean autoCreateTopics;
+
+    /**
+     * Whether topics should be automatically deleted when inactive, or rather or rather explicitly deleted through
```
[the remark this answers is not in the document]
Comment (ciaocloud) [reply pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/TopicLifecyclePolicies.java:40]: > I am confused about the relation between `InactiveTopicPolicies` and `TopicLifecycle`. Can you bring more details?

Both `InactiveTopicPolicies` and `TopicLifecycle` are fields declared in `Policies`. 
The `InactiveTopicPolicies` has a boolean value `deleteWhileInactive` which determines whether topics should be automatically deleted when inactive. 
We ask it to be same as that defined in `TopicLifecycle`, while `TopicLifecycle` also defines whether topics should be automatically created. 
This endpoint is used during tenant creation time.

(Also would update the above in the PR description)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 3]
```diff
@@ -2485,6 +2493,13 @@ private boolean shouldTopicBeRetained() {
         //If the topic-level policy already exists, the namespace-level policy cannot override the topic-level policy.
         Optional<TopicPolicies> topicPolicies = getTopicPolicies();
 
+        if (data.topicLifecycle != null) {
```
[the remark this answers is not in the document]
Comment (ciaocloud) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2496]: ack. thanks.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ciaocloud) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2498]: it's about topic creation/deletion, so we should `updateNamespaceValue`, right?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ciaocloud) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2496]: We can definitely refactor it into the `updateTopicPolicyByNamespacePolicy` method in `AbstractTopic`, but as that method would be used by both `PersistentTopic` and `NonPersistentTopic`, not sure it is the right thing to do for the non-persistent one to update `inactiveTopicPolices` with `topicLifecycle` info. will need to double-check.

File: pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/TopicLifecyclePolicies.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jason918) [reply pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/TopicLifecyclePolicies.java]: > We ask it to be same as that defined in `TopicLifecycle`


How can you guarantee this? Current code clearly missed some part.

And I don't think it's a good design that we have two field defines the same things ( `InactiveTopicPolicies#deleteWhileInactive` and `TopicLifecyclePolicies#autoDeleteTopics` )

**Comment by tisonkun** [comment]
Closed as stale and conflict. Please rebase and resubmit the patch if it's still relevant.

**Pull request closed without merging**